doc/man/man1/openchangepfadmin.1
.\" OpenChange Project Tools Man Pages
.\"
.\" This manpage is Copyright (C) 2007 Julien Kerihuel;
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one.
.\" 
.\" Since the OpenChange and Samba4 libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" Process this file with
.\" groff -man -Tascii openchangepfadmin.1
.\"
.TH OPENCHANGEPFADMIN 1 2013-01-24 "OpenChange 2.0 QUADRANT" "OpenChange Users' Manual"

.SH NAME
openchangepfadmin \- Exchange users and Public Folder administration tool

.SH SYNOPSIS
.nf
openchangepfadmin [-?] [-?|--help] [--usage] [-f|--database PATH] [-p|--profile PROFILE]
    [-P|--password PASSWORD] [--apassword=PASSWORD] [--adesc=DESCRIPTION] [--acomment=COMMENT]
    [--afullname=NAME] [--list] [--mkdir] [--rmdir] [--comment=COMMENT] [--dirclass=CLASS]
    [--adduser=USERNAME] [--rmuser=USERNAME] [--addright=RIGHT] [--rmright] [--modright=RIGHT]
    [--debuglevel=LEVEL] [--dump-data] [--folder=FOLDER] [--username=USERNAME]
.fi

.SH DESCRIPTION
openchangepfadmin is a administrative command line tool designed to
facilitate user management related operations (create, delete, modify)
on a remote Exchange server and manage public folder store and
permissions.

.SH COMMANDS

.TP
.B --adduser=USERNAME
Create a Exchange user with the username specified by USERNAME

.TP
.B --rmuser=USERNAME
Delete the Exchange user account specified by USERNAME

.TP
.B --list
List Public Folder hierarchy (IPM_SUBTREE)

.TP
.B --mkdir
Create a folder in the Public Folder store

.TP
.B --rmdir
Delete a folder in the Public Folder store

.TP
.B --addright=RIGHT
Add permission and roles for a user on a Public Folder directory. 

.TP
.B --modright=RIGHT
Modify permissions and roles for a user on a Public Folder directory

.TP
.B --rmright
Delete permissions and roles for a specific user


.SH OPTIONS

.TP
.B --database=DATABASE
.TP
.B -f
Set the MAPI profile database. If no database is specified, then
openchangeclient tries to load the default one:
.B $HOME/.openchange/profiles.ldb

.TP
.B --profile=PROFILE
.TP
.B -p
Set the profile to use. If a profile is not specified, and one of the
profiles has been set as the default in the profile database (for
example, using mapiprofile -S), then that default profile will be
used.

.TP
.B --password
.TP
.B -P
Set the password for the profile to use. This can be omitted if the
password is stored in the profile.

.TP
.B --apassword=PASSWORD
Define the password to set for the user specified with --adduser. If
no password is explicitely supplied, openchangepfadmin will arbitrary
set a random one.

.TP
.B --adesc=DESCRIPTION
This command can only be used with --adduser and specifies a
description for this account

.TP
.B --acomment=COMMENT
This command can only be used with --adduser and specifies a comment
for this account.

.TP
.B --afullname=NAME
This command can only be used with --adduser and specifies the user
full name for this account.

.TP
.B --comment=COMMENT
This command can only be used with --mkdir and specifies a comment for
the folder.

.TP
.B --dirclass=CLASS
This command can only be used with --mkdir and specifies the container
class of the directory we want to create. Possible values are:
IPF.Appointment, IPF.Contact, IPF.Journal, IPF.Note, IPF.StickyNote,
IPF.Task, IPF.Post

.TP
.B --folder=FOLDER
This command can only be used with --addright, --modright and
--rmright. Specify the folder where permissions and roles have to be
changed.

.TP
.B --username=USERNAME
This command can only be used with --addright, --modright and
--rmright. Specify the username we want to change permissions and
roles for.

.TP
.B --dump-data
Dump the hex data. This is only required for debugging or educational purposes.

.TP
.B --debuglevel LEVEL
.TP
.B -d
Set the debug level.

.SH EXAMPLES

.B Creating user
.nf
openchangepfadmin --adduser=linuxowner --apassword=linuxowner   \\
                    --adesc="Linux Owner Test account"          \\
		    --afullname="Linux Owner"
    mapiadmin_user_add       : MAPI_E_SUCCESS (0x0)
username: linuxowner
password: linuxowner
.fi

Creates a user account with username and password set to linuxowner.

.B Deleting user:
.nf
openchangepfadmin --rmuser=linuxowner
    mapiadmin_user_del       : MAPI_E_SUCCESS (0x0)
.fi

.B Create Public Folder:
.nf
openchangepfadmin --mkdir --folder=public_events --dirclass=IPF.Appointment
.fi

Creates a folder in the Public Folder store named public_events with a
container class set to Appointment. This folder will display calendar
items.

.B Delete Public Folder:
.nf
openchangepfadmin --rmdir --folder=public_events
.fi

.B List Public Folder hierarchy:
.nf
openchangepfadmin --list
.fi

.B Add Permission:
.nf
openchangepfadmin --username=linuxowner --folder=public_events \\
                    --addright=RoleOwner
Permission RoleOwner added for linuxowner on folder public_events
.fi

.B Modify Permission
.nf
openchangepfadmin --username=Anonymous --folder=public_events \\
  		    --modright=RoleNone
Permission changed to RoleNone for Anonymous on folder appointment
.fi

Note that you can only change permissions for a user already listed in
the ACL table.

.SH NOTE
Permissions and Roles possible values are:
.IP "\(bu" 2
RightsNone                    
.IP "\(bu" 2
RightsReadItems               
.IP "\(bu" 2
RightsCreateItems             
.IP "\(bu" 2
RightsEditOwn                 
.IP "\(bu" 2
RightsDeleteOwn               
.IP "\(bu" 2
RightsEditAll
.IP "\(bu" 2    
RightsDeleteAll
.IP "\(bu" 2
RightsCreateSubfolders
.IP "\(bu" 2
RightsFolderOwner
.IP "\(bu" 2
RightsFolderContact
.IP "\(bu" 2
RoleNone
.IP "\(bu" 2
RoleReviewer
.IP "\(bu" 2
RoleContributor
.IP "\(bu" 2
RoleNoneditingAuthor
.IP "\(bu" 2
RoleAuthor
.IP "\(bu" 2
RoleEditor
.IP "\(bu" 2
RolePublishAuthor
.IP "\(bu" 2
RolePublishEditor
.IP "\(bu" 2
RightsAll
.IP "\(bu" 2
RoleOwner			

.SH SEE ALSO
AddUserPermission, ModifyUserPermission, RemoveUserPermission

.SH AUTHOR
Julien Kerihuel <j.kerihuel at openchange dot org>


ChangeLog
2010-08-29
    jkerihuel
	[r2144]
		Add CopyProfile and DuplicateProfile to IProfAdmin interface

2010-08-27
    jkerihuel
	[r2142]
		Add the localaddress patch prior Samba4 integration. Pending new
		Samba4 localaddress patch fixing memory leaks in the pipe.
	[r2141]
		Include a mapitest test suite / challenge for lzxpress
		
		The idea of this test is to check if lzxpress compressed blobs
		provided by OpenChange are the same than Outlook ones. The test takes
		a Outlook->Exchange capture file with compressed blob, pull the
		content, decompressed the content, recompress the content and check
		the output for differing bytes.
		
		This challenge code should give us good information when lzxpress
		compress code is fixed. Additionally you can uncomment the compression
		code in libmapi/emsmdb.c to do further testing.
	[r2140]
		- Add implementation for EcDoRpcExt2 (0xB) RPC opnum. Current code
		  doesn't provide compression for request and use obfuscation
		  instead. However mapi_response makes use of xpress and are
		  decompressed properly.
		
		- EcDoRpcExt2 implemented without subcontexts but uint8_t array
		
		- add a wrapper for emsmdb transaction to switch between
		  EcDoConnect/EcDoRpc and EcDoConnectEx/EcDoRpcExt2 depending on MAPI
		  profile 'exchange version' attribute.
		
		NOTE: emsmdb.c implements the logic for using compressed blob and
		checks whether compressed or obfuscation should be used depending on
		blob length. However this behavior has been commented out because
		comrpessed blobs are rejected by Exchange (ecRpcFormat 0x4B6 error).
2010-08-26
    jkerihuel
	[r2139]
		Reference profile structure in mapitest structure.
		
		Use profile->mailbox instead of szDisplayName for sending mapitest
		emails. This also fix ResolveNames mapitest issues when conflicts
		between usernames exist (e.g. user1, user11, user100).
    bradh
	[r2137]
		Prevent segfault if we don't have any appointments.
		
		Resolves issue #284.
		
		Also prevent leaking talloc context in a couple of error cases.
	[r2136]
		Merge Rev 2133 from branches/current-samba4 to trunk.
		
		I inadvertently committed this to the wrong branch.
		
		This pushes the fix for #229 into trunk.

2010-08-23
    jelmer
	[r2134]
		Update bzr ignores.

2010-08-18
    jkerihuel
	[r2131]
		Add a git update case in installsamba.sh script and make it available
		within OpenChange Makefile.

2010-08-17
    jkerihuel
	[r2127]
		Manually commit remaining changes from current-samba4 into trunk
	[r2126]
		Merge current-samba4 branch back into trunk

2010-08-11
    bradh
	[r2117]
		Microsoft confirmed that RopGetValidAttachments isn't supported in Exchange 2010.
	[r2114]
		fix missing part of r2113.
	[r2113]
		Fixes to better work with Exchange 2010.
		
		Still some open issues to resolving bug 227.
	[r2112]
		Fix handling of "not applicable" tests when using --mapi-calls
		
		Prior to this, if you asked for a test that was after a test that was
		flagged as suppressed, the test you asked for didn't get run, and the
		flagged test was reported instead.
2010-08-10
    bradh
	[r2111]
		Correctly handle cases where the start date of the range is after the first calendar appointment.
		
		Resolves bug:272.
		
		Thanks to Eric Reischer for reporting the problem and testing the fix.

2010-08-03
    bradh
	[r2107]
		trivial comment cleanup.

2010-08-01
    jkerihuel
	[r2105]
		Move from auto-generated mapicode.h to static code.
	[r2104]
		Add configure.ac support for gthread implementation but give preferences to pthread
		Add THREAD_CFLAGS to config.mk.in
		ifdef between pthread and gthread in libmapi_private.h
	[r2103]
		Check for libpthread in configure.ac and make this lib optional
		Add references to config.h in libmapi_private.h
		Replace references to config.h with libmapi_private.h

2010-07-31
    jkerihuel
	[r2101]
		- Add an exchange-version parameter to mapiprofile structure and tool
		  which let user choose between 0x0/0x2 or 0xA/0xB
		- Tweak libmapi implementation to allow EcDoConnectEx usage
		- Improve mapiprofile --dump to show more useful parameters
	[r2100]
		- Add custom NDR implementation for EcDoConnectEx to bypass pointers/array problems
		- make use of mapi2k7_AuxInfo structure for rgbAuxOut
		
		- Create a subcontext for AUX_HEADER structures to jump over unknown cases properly.
	[r2099]
		Handle the no obfuscation nor compression RPC_HEADER_EXT flags case 
    bradh
	[r2098]
		Trivial api documentation change.

2010-07-29
    bradh
	[r2095]
		Another trivial API documentation fix.
	[r2094]
		More trivial API documentation fixes.
	[r2093]
		Minor API documentation fix.
	[r2092]
		Implement some null pointer checks.
		
		Also rewords some API documentation.
		
		Original issue and patch by Erik Hovland - thanks again. I implemented it
		with some slight changes.

2010-07-28
    bradh
	[r2090]
		Fix RecipientRow IDL to match MS-OXCDATA 2.9.3
		
		This allows handling of addresses that contain X400 format content.
		
		Patch by Samuel Sieb and I, resolves ticket #250.

2010-07-26
    jkerihuel
	[r2087]
		Fix a crash in openchangeclient vs Exchange 2010:
		When Queryrows is called on Exchange 2010 and no rows are available, QueryRows returns MAPI_E_NOT_IMPLEMENTED error (0x8004FFFF).
		Ensure we have at least one message to fetch in openchangeclient --fetchmail prior calling QueryRows.

2010-07-24
    jkerihuel
	[r2079]
		Very basic use of OpenUserMailbox in openchangeclient.
		Doesn't handle clustered Exchange servers.
    bradh
	[r2077]
		Rename a few function arguments to avoid doxygen warnings.
		
		This is a workaround, but will allow easier checking for the future.
	[r2076]
		Update doxygen configuration to match latest version
	[r2075]
		Add missing doxygen parameter information.
	[r2074]
		Fix a couple of doxygen issues.

2010-07-23
    jkerihuel
	[r2072]
		Use sed in Makefile to replace local headers reference in installed headers
	[r2071]
		Use local headers rather than installed ones

utils/openchangepfadmin.c
/*
   Public Folders Administration Tool

   OpenChange Project

   Copyright (C) Julien Kerihuel 2007

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "libmapi/libmapi.h"
#include "libmapiadmin/libmapiadmin.h"
#include <popt.h>
#include <param.h>

#include "openchangepfadmin.h"
#include "openchange-tools.h"

static int32_t get_aclrights(const char *permission)
{
	uint32_t	i;

	if (!permission) return -1;

	for (i = 0; aclrights[i].name != NULL; i++) {
		if (!strcmp(permission, aclrights[i].name)) {
			return aclrights[i].value;
		}
	}
	return -1;
}

static void list_aclrights(void)
{
	uint32_t	i;

	printf("Valid permissions:\n");
	for (i = 0; aclrights[i].name != NULL; i++) {
		printf("\t%s\n", aclrights[i].name);
	}
}

static int32_t check_IPF_class(const char *dirclass)
{
	uint32_t	i;

	if (!dirclass) return -1;

	for (i = 0; IPF_class[i]; i++) {
		if (!strcmp(dirclass, IPF_class[i])) {
			return 0;
		}
	}

	return -1;
}

static void list_IPF_class(void)
{
	uint32_t	i;

	printf("Valid IPF Classes:\n");
	for (i = 0; IPF_class[i] != NULL; i++) {
		printf("\t%s\n", IPF_class[i]);
	}
}

static bool get_child_folders_pf(TALLOC_CTX *mem_ctx, mapi_object_t *parent, mapi_id_t folder_id, int count)
{
	enum MAPISTATUS		retval;
	bool			ret;
	mapi_object_t		obj_folder;
	mapi_object_t		obj_htable;
	struct SPropTagArray	*SPropTagArray;
	struct SRowSet		rowset;
	const char	       	*name;
	const uint32_t		*child;
	uint32_t		index;
	const uint64_t		*fid;
	int			i;

	mapi_object_init(&obj_folder);
	retval = OpenFolder(parent, folder_id, &obj_folder);
	if (retval != MAPI_E_SUCCESS) return false;

	mapi_object_init(&obj_htable);
	retval = GetHierarchyTable(&obj_folder, &obj_htable, 0, NULL);
	if (retval != MAPI_E_SUCCESS) return false;

	SPropTagArray = set_SPropTagArray(mem_ctx, 0x3,
					  PR_DISPLAY_NAME_UNICODE,
					  PR_FID,
					  PR_FOLDER_CHILD_COUNT);
	retval = SetColumns(&obj_htable, SPropTagArray);
	MAPIFreeBuffer(SPropTagArray);
	if (retval != MAPI_E_SUCCESS) return false;
	
	while (((retval = QueryRows(&obj_htable, 0x32, TBL_ADVANCE, &rowset)) != MAPI_E_NOT_FOUND) && rowset.cRows) {
		for (index = 0; index < rowset.cRows; index++) {
			fid = (const uint64_t *)find_SPropValue_data(&rowset.aRow[index], PR_FID);
			name = (const char *)find_SPropValue_data(&rowset.aRow[index], PR_DISPLAY_NAME_UNICODE);
			child = (const uint32_t *)find_SPropValue_data(&rowset.aRow[index], PR_FOLDER_CHILD_COUNT);

			for (i = 0; i < count; i++) {
				printf("|   ");
			}
			printf("|---+ %-15s\n", name);
			if (*child) {
				ret = get_child_folders_pf(mem_ctx, &obj_folder, *fid, count + 1);
				if (ret == false) return ret;
			}
			
		}
	}
	return true;
}

static enum MAPISTATUS openchangepfadmin_getdir(TALLOC_CTX *mem_ctx, 
						mapi_object_t *obj_container,
						mapi_object_t *obj_child,
						const char *folder)
{
	enum MAPISTATUS		retval;
	struct SPropTagArray	*SPropTagArray;
	struct SRowSet		rowset;
	mapi_object_t		obj_htable;
	const char		*name;
	const uint64_t		*fid;
	uint32_t		index;

	mapi_object_init(&obj_htable);
	retval = GetHierarchyTable(obj_container, &obj_htable, 0, NULL);
	MAPI_RETVAL_IF(retval, GetLastError(), NULL);

	SPropTagArray = set_SPropTagArray(mem_ctx, 0x2,
					  PR_DISPLAY_NAME_UNICODE,
					  PR_FID);
	retval = SetColumns(&obj_htable, SPropTagArray);
	MAPIFreeBuffer(SPropTagArray);
	if (retval != MAPI_E_SUCCESS) return MAPI_E_NOT_FOUND;
	
	while (((retval = QueryRows(&obj_htable, 0x32, TBL_ADVANCE, &rowset)) != MAPI_E_NOT_FOUND) && rowset.cRows) {
		for (index = 0; index < rowset.cRows; index++) {
			fid = (const uint64_t *)find_SPropValue_data(&rowset.aRow[index], PR_FID);
			name = (const char *)find_SPropValue_data(&rowset.aRow[index], PR_DISPLAY_NAME_UNICODE);

			if (name && fid && !strcmp(name, folder)) {
				retval = OpenFolder(obj_container, *fid, obj_child);
				MAPI_RETVAL_IF(retval, GetLastError(), NULL);
				return MAPI_E_SUCCESS;
			}
		}
	}

	errno = MAPI_E_NOT_FOUND;
	return MAPI_E_NOT_FOUND;
}

static enum MAPISTATUS openchangepfadmin_mkdir(mapi_object_t *obj_container, 
					       const char *name, 
					       const char *comment, 
					       const char *type)
{
	enum MAPISTATUS		retval;
	mapi_object_t		obj_folder;
	struct SPropValue	props[1];
	uint32_t		prop_count = 0;

	mapi_object_init(&obj_folder);
	retval = CreateFolder(obj_container, FOLDER_GENERIC, name, comment, OPEN_IF_EXISTS, &obj_folder);
	MAPI_RETVAL_IF(retval, GetLastError(), NULL);

	if (type) {
		set_SPropValue_proptag(&props[0], PR_CONTAINER_CLASS, (const void *) type);
		prop_count++;

		retval = SetProps(&obj_folder, 0, props, prop_count);
		mapi_object_release(&obj_folder);
		MAPI_RETVAL_IF(retval, GetLastError(), NULL);
	}
	
	return MAPI_E_SUCCESS;
}

static enum MAPISTATUS openchangepfadmin_rmdir(TALLOC_CTX *mem_ctx,
					       mapi_object_t *obj_container,
					       const char *name)
{
	enum MAPISTATUS		retval;
	mapi_object_t		obj_folder;

	mapi_object_init(&obj_folder);

	retval = openchangepfadmin_getdir(mem_ctx, obj_container, &obj_folder, name);
	MAPI_RETVAL_IF(retval, GetLastError(), NULL);

	retval = EmptyFolder(&obj_folder);
	MAPI_RETVAL_IF(retval, GetLastError(), NULL);

	retval = DeleteFolder(obj_container, mapi_object_get_id(&obj_folder),
			      DEL_FOLDERS, NULL);
	MAPI_RETVAL_IF(retval, GetLastError(), NULL);

	mapi_object_release(&obj_folder);

	return MAPI_E_SUCCESS;
}

int main(int argc, const char *argv[])
{
	TALLOC_CTX		*mem_ctx;
	enum MAPISTATUS		retval;
	struct mapi_context	*mapi_ctx;
	struct mapi_session	*session = NULL;
	mapi_object_t		obj_store;
	mapi_object_t		obj_ipm_subtree;
	mapi_object_t		obj_folder;
	mapi_id_t		id_pf;
	poptContext		pc;
	int			opt;
	const char		*opt_profdb = NULL;
	char			*opt_profname = NULL;
	const char		*opt_password = NULL;
	const char		*opt_comment = NULL;
	const char		*opt_dirclass = NULL;
	const char		*opt_adduser = NULL;
	const char		*opt_rmuser = NULL;
	const char		*opt_apassword = NULL;
	const char		*opt_adesc = NULL;
	const char		*opt_acomment = NULL;
	const char		*opt_afullname = NULL;
	const char		*opt_addright = NULL;
	bool			opt_rmright = false;
	const char		*opt_modright = NULL;
	const char		*opt_folder = NULL;
	const char		*opt_debug = NULL;
	const char		*opt_username = NULL;	
	int32_t			opt_permission = -1;
	bool			opt_ipm_list = false;
	bool			opt_mkdir = false;
	bool			opt_rmdir = false;
	bool			opt_dumpdata = false;

	enum {OPT_PROFILE_DB=1000, OPT_PROFILE, OPT_PASSWORD, OPT_IPM_LIST, 
	      OPT_MKDIR, OPT_RMDIR, OPT_COMMENT, OPT_DIRCLASS, OPT_ACL, 
	      OPT_ADDUSER, OPT_RMUSER, OPT_DEBUG, OPT_APASSWORD, OPT_ADESC, 
	      OPT_ACOMMENT, OPT_AFULLNAME, OPT_ADDRIGHT, OPT_RMRIGHT, 
	      OPT_MODRIGHT, OPT_USERNAME, OPT_FOLDER, OPT_DUMPDATA};

	struct poptOption long_options[] = {
		POPT_AUTOHELP
		{"database", 'f', POPT_ARG_STRING, NULL, OPT_PROFILE_DB, "set the profile database path", "PATH"},
		{"profile", 'p', POPT_ARG_STRING, NULL, OPT_PROFILE, "set the profile name", "PROFILE"},
		{"password", 'P', POPT_ARG_STRING, NULL, OPT_PASSWORD, "set the profile password", "PASSWORD"},
		{"apassword", 0, POPT_ARG_STRING, NULL, OPT_APASSWORD, "set the account password", "PASSWORD"},
		{"adesc", 0, POPT_ARG_STRING, NULL, OPT_ADESC, "set the account description", "DESCRIPTION"},
		{"acomment", 0, POPT_ARG_STRING, NULL, OPT_ACOMMENT, "set the account comment", "COMMENT"},
		{"afullname", 0, POPT_ARG_STRING, NULL, OPT_AFULLNAME, "set the account full name", "NAME"},
		{"list",0, POPT_ARG_NONE, NULL, OPT_IPM_LIST, "list IPM subtree directories", NULL},
		{"mkdir", 0, POPT_ARG_NONE, NULL, OPT_MKDIR, "create a Public Folder directory", NULL},
		{"rmdir", 0, POPT_ARG_NONE, NULL, OPT_RMDIR, "delete a Public Folder directory", NULL},
		{"comment", 0, POPT_ARG_STRING, NULL, OPT_COMMENT, "set the folder comment", "COMMENT"},
		{"dirclass", 0, POPT_ARG_STRING, NULL, OPT_DIRCLASS, "set the folder class", "CLASS"},
		{"adduser", 0, POPT_ARG_STRING, NULL, OPT_ADDUSER, "add Exchange user", "USERNAME"},
		{"rmuser", 0, POPT_ARG_STRING, NULL, OPT_RMUSER, "delete Exchange user", "USERNAME"},
		{"addright", 0, POPT_ARG_STRING, NULL, OPT_ADDRIGHT, "add MAPI permissions to PF folder", "RIGHT"},
		{"rmright", 0, POPT_ARG_NONE, NULL, OPT_RMRIGHT, "remove MAPI permissions to PF folder", NULL},
		{"modright", 0, POPT_ARG_STRING, NULL, OPT_MODRIGHT, "modify MAPI permissions to PF folder", "RIGHT"},
		{"debuglevel", 0, POPT_ARG_STRING, NULL, OPT_DEBUG, "set debug level", "LEVEL"},
		{"dump-data", 0, POPT_ARG_NONE, NULL, OPT_DUMPDATA, "Dump the hex data", NULL},
		{"folder", 0, POPT_ARG_STRING, NULL, OPT_FOLDER, "specify the Public Folder directory", "FOLDER"},
		{"username", 0, POPT_ARG_STRING, NULL, OPT_USERNAME, "specify the username to use", "USERNAME"},
		POPT_OPENCHANGE_VERSION
		{ NULL, 0, POPT_ARG_NONE, NULL, 0, NULL, NULL }
	};

	mem_ctx = talloc_named(NULL, 0, "openchangepfadmin");

	pc = poptGetContext("openchangepfadmin", argc, argv, long_options, 0);

	while ((opt = poptGetNextOpt(pc)) != -1) {
		switch (opt) {
		case OPT_DEBUG:
			opt_debug = poptGetOptArg(pc);
			break;
		case OPT_DUMPDATA:
			opt_dumpdata = true;
			break;
		case OPT_PROFILE_DB:
			opt_profdb = poptGetOptArg(pc);
			break;
		case OPT_PROFILE:
			opt_profname = talloc_strdup(mem_ctx, poptGetOptArg(pc));
			break;
		case OPT_PASSWORD:
			opt_password = poptGetOptArg(pc);
			break;
		case OPT_IPM_LIST:
			opt_ipm_list = true;
			break;
		case OPT_MKDIR:
			opt_mkdir = true;
			break;
		case OPT_RMDIR:
			opt_rmdir = true;
			break;
		case OPT_COMMENT:
			opt_comment = poptGetOptArg(pc);
			break;
		case OPT_DIRCLASS:
			opt_dirclass = poptGetOptArg(pc);
			break;
		case OPT_ADDUSER:
			opt_adduser = poptGetOptArg(pc);
			break;
		case OPT_RMUSER:
			opt_rmuser = poptGetOptArg(pc);
			break;
		case OPT_APASSWORD:
			opt_apassword = poptGetOptArg(pc);
			break;
		case OPT_ADESC:
			opt_adesc = poptGetOptArg(pc);
			break;
		case OPT_ACOMMENT:
			opt_acomment = poptGetOptArg(pc);
			break;
		case OPT_AFULLNAME:
			opt_afullname = poptGetOptArg(pc);
			break;
		case OPT_ADDRIGHT:
			opt_addright = poptGetOptArg(pc);
			break;
		case OPT_RMRIGHT:
			opt_rmright = true;
			break;
		case OPT_MODRIGHT:
			opt_modright = poptGetOptArg(pc);
			break;
		case OPT_FOLDER:
			opt_folder = poptGetOptArg(pc);
			break;
		case OPT_USERNAME:
			opt_username = poptGetOptArg(pc);
			break;
		default:
			break;
		};
	}

	/* Sanity check on options */

	if ((opt_mkdir == true && !opt_folder) || (opt_rmdir == true && !opt_folder) ||
	    (opt_addright && !opt_folder) || (opt_rmright == true && !opt_folder) || 
	    (opt_modright && !opt_folder)) {
		printf("You need to specify a directory with --folder option\n");
		return (-1);
	}

	if ((opt_addright && !opt_username) || (opt_rmright == true && !opt_username) ||
	    (opt_modright && !opt_username)) {
		printf("You need to specify a username with --username option\n");
		return (-1);
	}

	/* dirclass sanity check */
	if (opt_dirclass) {
		if (check_IPF_class(opt_dirclass) == -1) {
			list_IPF_class();
			return (-1);
		}
	}

	if (!opt_profdb) {
		opt_profdb = talloc_asprintf(mem_ctx, DEFAULT_PROFDB, getenv("HOME"));
	}

	/**
	 * Initialize MAPI subsystem
	 */
	retval = MAPIInitialize(&mapi_ctx, opt_profdb);
	if (retval != MAPI_E_SUCCESS) {
		mapi_errstr("MAPIInitialize", GetLastError());
		exit (1);
	}

	/* debug options */
	SetMAPIDumpData(mapi_ctx, opt_dumpdata);

	if (opt_debug) {
		SetMAPIDebugLevel(mapi_ctx, atoi(opt_debug));
	}

	/**
	 * If no profile is specified, try to load the default one
	 * from the database
	 */
	if (!opt_profname) {
		retval = GetDefaultProfile(mapi_ctx, &opt_profname);
		if (retval != MAPI_E_SUCCESS) {
			mapi_errstr("GetDefaultProfile", GetLastError());
			exit (1);
		}
	}

	retval = MapiLogonEx(mapi_ctx, &session, opt_profname, opt_password);
	talloc_free(opt_profname);
	if (retval != MAPI_E_SUCCESS) {
		mapi_errstr("MapiLogonEx", GetLastError());
		exit (1);
	}

	/**
	 * User management operations
	 */

	if (opt_adduser) {
		struct mapiadmin_ctx	*mapiadmin_ctx;

		mapiadmin_ctx = mapiadmin_init(session);
		mapiadmin_ctx->username = opt_adduser;
		mapiadmin_ctx->password = opt_apassword;
		mapiadmin_ctx->description = opt_adesc;
		mapiadmin_ctx->comment = opt_acomment;
		mapiadmin_ctx->fullname = opt_afullname;

		retval = mapiadmin_user_add(mapiadmin_ctx);
		mapi_errstr("mapiadmin_user_add", GetLastError());
		if (retval != MAPI_E_SUCCESS) {
			exit (1);
		}
		printf("username: %s\n", mapiadmin_ctx->username);
		printf("password: %s\n", mapiadmin_ctx->password);
	}

	if (opt_rmuser) {
		struct mapiadmin_ctx *mapiadmin_ctx;
		
		mapiadmin_ctx = mapiadmin_init(session);
		mapiadmin_ctx->username = opt_rmuser;

		retval = mapiadmin_user_del(mapiadmin_ctx);
		mapi_errstr("mapiadmin_user_del", GetLastError());
	}

	/**
	 * Open Public Folder Store
	 */

	mapi_object_init(&obj_store);
	retval = OpenPublicFolder(session, &obj_store);
	if (retval != MAPI_E_SUCCESS) {
		mapi_errstr("OpenPublicFolder", GetLastError());
		goto end;
	}

	/* Open IPM Subtree (All Public Folders) */
	retval = GetDefaultPublicFolder(&obj_store, &id_pf, olFolderPublicIPMSubtree);
	if (retval != MAPI_E_SUCCESS) {
		goto end;
	}

	mapi_object_init(&obj_ipm_subtree);
	retval = OpenFolder(&obj_store, id_pf, &obj_ipm_subtree);
	if (retval != MAPI_E_SUCCESS) {
		mapi_errstr("OpenFolder", GetLastError());
		goto end;
	}

	/* create directory */
	if (opt_mkdir == true) {
		retval = openchangepfadmin_mkdir(&obj_ipm_subtree, opt_folder, opt_comment, opt_dirclass);
		if (retval != MAPI_E_SUCCESS) {
			mapi_errstr("mkdir", GetLastError());
			goto end;
		}
	}

	/* remove directory */
	if (opt_rmdir == true) {
		retval = openchangepfadmin_rmdir(mem_ctx, &obj_ipm_subtree, opt_folder);
		if (retval != MAPI_E_SUCCESS) {
			mapi_errstr("rmdir", GetLastError());
			goto end;
		}
		
	}

	/* add permissions */
	if (opt_addright) {
		if ((opt_permission = get_aclrights(opt_addright)) == -1) {
			printf("Invalid permission\n");
			list_aclrights();
			goto end;
		} else {
			mapi_object_init(&obj_folder);
			retval = openchangepfadmin_getdir(mem_ctx, &obj_ipm_subtree, &obj_folder, opt_folder);
			if (retval != MAPI_E_SUCCESS) {
				printf("Invalid directory\n");
				goto end;
			}
			retval = AddUserPermission(&obj_folder, opt_username, opt_permission);
			if (retval != MAPI_E_SUCCESS) {
				mapi_errstr("AddUserPermission", GetLastError());
				mapi_object_release(&obj_folder);
				goto end;
			}
			mapi_object_release(&obj_folder);
		}
		printf("Permission %s added for %s on folder %s\n", opt_addright, opt_username, opt_folder);
	}

	/* remove permissions */
	if (opt_rmright == true) {
		mapi_object_init(&obj_folder);
		retval = openchangepfadmin_getdir(mem_ctx, &obj_ipm_subtree, &obj_folder, opt_folder);
		if (retval != MAPI_E_SUCCESS) {
			printf("Invalid directory\n");
			goto end;
		}
		retval = RemoveUserPermission(&obj_folder, opt_username);
		if (retval != MAPI_E_SUCCESS) {
			mapi_errstr("RemoveUserPermission", GetLastError());
			mapi_object_release(&obj_folder);
			goto end;
		}
		mapi_object_release(&obj_folder);
		printf("Permission removed for %s on folder %s\n", opt_username, opt_folder);
	}

	/* modify permissions */
	if (opt_modright) {
		opt_permission = get_aclrights(opt_modright);
		if (opt_permission == -1) {
			printf("Invalid permission\n");
			list_aclrights();
			goto end;
		} else {
			mapi_object_init(&obj_folder);
			retval = openchangepfadmin_getdir(mem_ctx, &obj_ipm_subtree, &obj_folder, opt_folder);
			if (retval != MAPI_E_SUCCESS) {
				printf("Invalid directory\n");
				goto end;
			}
			retval = ModifyUserPermission(&obj_folder, opt_username, opt_permission);
			if (retval != MAPI_E_SUCCESS) {
				mapi_errstr("ModifyUserPermission", GetLastError());
				mapi_object_release(&obj_folder);
				goto end;
			}
			mapi_object_release(&obj_folder);
		}
		printf("Permission changed to %s for %s on folder %s\n", opt_modright, opt_username, opt_folder);
	}

	/* list directories */
	if (opt_ipm_list == true) {
		printf("+ All Public Folders\n");
		get_child_folders_pf(mem_ctx, &obj_ipm_subtree, id_pf, 0);
	}

	/**
	 * Uninitialize MAPI subsystem
	 */
end:
	mapi_object_release(&obj_store);

	MAPIUninitialize(mapi_ctx);
	talloc_free(mem_ctx);
	return 0;
}
